x-pack/plugins/observability/server/services/slo/resource_installer.ts

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  ClusterPutComponentTemplateRequest,
+  IndicesPutIndexTemplateRequest,
+} from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import type { ElasticsearchClient, Logger } from '@kbn/core/server';
+
+import {
+  SLO_COMPONENT_TEMPLATE_MAPPINGS_NAME,
+  SLO_COMPONENT_TEMPLATE_SETTINGS_NAME,
+  SLO_INDEX_TEMPLATE_NAME,
+} from '../../assets/common';
+import { sloMappingsTemplate } from '../../assets/component_templates/slo_mappings_template';
+import { sloSettingsTemplate } from '../../assets/component_templates/slo_settings_template';
+import { sloIndexTemplate } from '../../assets/index_templates/slo_index_templates';
+
+const DEFAULT_INSTALLATION_TIMEOUT = 60 * 1000; // 1 minute
+
+interface ResourceInstallerConfig {
+  installationTimeout: number;
+}
+
+export class ResourceInstaller {
+  constructor(
+    private esClient: ElasticsearchClient,
+    private logger: Logger,
+    private config: ResourceInstallerConfig = { installationTimeout: DEFAULT_INSTALLATION_TIMEOUT }
+  ) {}
+
+  public async ensureCommonResourcesInstalled(): Promise<void> {
+    const existingIndexTemplate = await this.checkIndexTemplateExists();
+
+    if (existingIndexTemplate) {
+      this.logger.debug(`Skipping installation of resources shared between all SLO indices`);
+      return;
+    }
+
+    await this.withTimeout('resources shared between all SLO indices', async () => {
+      try {
+        await Promise.all([
+          this.createOrUpdateComponentTemplate({
+            name: SLO_COMPONENT_TEMPLATE_MAPPINGS_NAME,
+            ...sloMappingsTemplate,
+          }),
+
+          this.createOrUpdateComponentTemplate({
+            name: SLO_COMPONENT_TEMPLATE_SETTINGS_NAME,
+            ...sloSettingsTemplate,
+          }),
+        ]);
+
+        await this.createOrUpdateIndexTemplate({
+          name: SLO_INDEX_TEMPLATE_NAME,
+          ...sloIndexTemplate,
+        });
+      } catch (err) {
+        this.logger.error(`Error installing resources shared for SLO - ${err.message}`);
+        throw err;
+      }
+    });
+  }
+
+  private async withTimeout(details: string, installer: () => Promise<void>): Promise<void> {
```

Comment:
I wonder how`withTimeout` function could scale and work with different machine sizes. 
I believe the tricky part is to define `installationTimeout` value because this would vary a lot based on the Kibana hosting machine and its resources and the SLO index creation time.

@kdelemme, could you please elaborate more about what you are trying to achieve using `withTimeout` ? the ES client by default has its own HTTP timeout, isn't it enough to rely on that?

Reply:
When discussing this ticket with @simianhacker we decided to inspire ourself from the work done in [the rule registry plugin](https://github.com/kdelemme/kibana/blob/feat/139414-route-slo/x-pack/plugins/rule_registry/server/rule_data_plugin_service/resource_installer.ts#L38)

That being said, since we are not using this ResourceInstaller at the Plugin Installation step, I don't see why we could not just leverage the esClient timeout and simplify the whole code here. Good idea